### PR TITLE
Possible source of leaked delayable writables 

### PR DESCRIPTION
**Pull request by not-napoleon**
In a couple of error paths, it's possible that `QueryPhaseResultsConsumer#PendingMerges` might not have consumed all the aggregations it's trying to merge.  It is never the less important to release those aggregations so that we don't leak memory or hold references to them.  This PR achieves that by using the `Releasables.close()` mechanism, which will execute each close action, even if earlier actions had exceptions.  This ensures that all of the aggregations get released and that the circuit breaker gets cleaned up.

ORIGINAL DESCRIPTION:
@DaveCTurner So, I haven't written a test for it yet, but I saw a path that might cause the leak we were looking for.  In `QueryPhaseResultConsumer`, we iterate all the `QuerySearchResult` instances and call `consumeAggs` on them.  But, `consumeAggs` can throw an unchecked exception if the aggregation fails to deserialize correctly (which definitely can happen, especially with pipeline aggs).  In that case, any unconsumed aggs would leak, I think.

While I was touching it, I turned a couple of related `asserts` into `exceptions`.  I don't think either is expensive to evaluate, and I'd rather know we don't do that in production than know we didn't write a test case that does that.

/cc @nik9000

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java:264]: Should we do this in all circumstances (i.e. in a `finally` block)?

(also it's ok to pass nulls to `Releasables#close`)

**Review by DaveCTurner — COMMENTED** [review]
I left a question

File: server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java:264]: I'd be fine with as much paranoia as you could manage. I'd probably move this to the top of the method to be honest.

I hadn't realized the aggregations were closeable - should we call `releaseAggs` anyway? That nulls the reference too I think.

File: server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java:268]: Suggest avoiding the intermediate list altogether (unfortunately type inference isn't clever enough to do this on its own)

```suggestion
                Releasables.close(
                    (Iterable<DelayableWriteable<InternalAggregations>>) (() -> buffer.stream()
                        .map(QuerySearchResult::aggregations)
                        .iterator())
                );
```

**Review by DaveCTurner — COMMENTED** [review]
This has passed many hundreds of iterations of the test for me so I have no objections, but I'd like someone more familiar with this area of the code to have the final say.

I left one small (non-blocking) suggestion.

**Review by henningandersen — COMMENTED** [review]
One small comment.

**Review by nik9000 — APPROVED** [review]
It's ok with me. It's super paranoid. I like paranoid, especially here.

**Comment by not-napoleon** [comment]
@jimczi Would appreciate it if you could give this a once over before I merge it.

File: server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java:258]: This condition should move to the `finally` block. Something like:
```
if (hasFailure()) {
  assert circuitBreakerBytes == 0;
} else {
  assert circuitBreakerBytes >= 0;
  circuitBreaker.addWithoutBreaking(-circuitBreakerBytes);
  circuitBreakerBytes = 0;
}
```

**Review by jimczi — CHANGES_REQUESTED** [review]
Thanks @not-napoleon, I left one comment

**Comment by not-napoleon** [comment]
After addressing Jim's comment, all of the functionality was in the finally block except for the `hasPendingMerges()` check.  At that point, it seemed more logical to just move the `hasPendingMerges()` check to the end of the method and remove the try-finally entirely.

**Review by nik9000 — APPROVED** [review]
LGTM. Nice and paranoid.

**Review by jimczi — APPROVED** [review]
LGTM

**Comment by jgq2008303393** [comment]
From which version does the impact begin? Will other versions backport this bugfix?

**Comment by not-napoleon** [comment]
@jgq2008303393 This has been backported to 7.15.2 and 7.16.0.  There are no plans for further backports.

As for impacted versions, that question is best suited to [the forums](https://discuss.elastic.co/).

**Pull request merged**
